datastore/adapter/keytransform.py

```diff
@@ -338,3 +341,204 @@ class ObjectNestedPathAdapter(
+		return datastore.Key(str(key).rjust(length, "_")[:length])
+
+	@staticmethod
+	def _next_to_last(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length + 1, "_")[:length])
+
+	@classmethod
+	async def _parse_sharding_function(
+			cls: typing.Type[T],
+			child_datastore: DS,
+			sharding_key: datastore.Key,
+			default_sharding_func: str
+	) -> KEY_TRANSFORM_T:
+		"""Determine the sharding function from the sharding_key.
+		
+		Arguments
+		---------
+		cls
+			Class variable passed in due to `classmethod`
+		child_datastore
+			Underlying datastore always used to access the `sharding_key`
+		sharding_key
+			Key used to access the sharding function in the `child_datastore`
+		default_sharding_func
+			Default sharding func to use if one is not found in `child_datastore`
+
+		Raises
+		------
+		Exception
+			Empty shard identifier file.
+		Exception
+			Prefix was not present in sharding file
+		Exception
+			Sharding file was in incorrect format
+		Exception
+			Expect `v1` format
+		Exception
+			`key_length` shouold be an integer
+		Exception
+			`function_name` was not `prefix`, `suffix`, or `next-to-last`
+		RuntimeError
+			An internal error occurred
+		"""
+
+		prefix: str = cls.prefix  # type: ignore[attr-defined] # noqa: F821
+
+		sharding_func: str
+		need_to_store_sharding_func: bool = False
+		try:
+			# Try reading existing sharding func used
+			sharding_func = next(   # type: ignore[assignment] # noqa: F821
+				iter(await child_datastore.get_all(sharding_key))
+			)
+		except KeyError:
+			# Use the given default sharding func
+			sharding_func = default_sharding_func
+			need_to_store_sharding_func = True
+
+		# Ensure not empty
+		if not sharding_func:
+			raise Exception("Empty shard identifier file.")
+
+		sharding_func = sharding_func.strip()
+
+		# Returns a 3-tuple containing the part before the separator,
+		# the separator itself, and the part after the separator
+		_, _, sharding_fn = sharding_func.partition(prefix)
+
+		# Ensure proper format
+		if not sharding_fn:
+			raise Exception(f"Prefix ({prefix}) was not present in {sharding_func}")
+
+		parts: list = sharding_fn.split("/")
+
+		if len(parts) != 3:
+			raise Exception(
+				f"invalid shard identifier: {sharding_fn}.\n"
+				f"Expecting form: {prefix}/version/function_name/key_length"
+			)
+
+		version, function_name, length = parts
+
+		if version != "v1":
+			raise Exception(f"Expected 'v1' for version string got: {version}\n")
+
+		try:
+			length = int(length)
+		except ValueError:
+			raise Exception(f"Invalid parameter: {length}. Should be integer representing `key_length`.")
+
+		funcs: dict = {
+			"prefix": cls._prefix,  # type: ignore[attr-defined] # noqa: F821
+			"suffix": cls._suffix,  # type: ignore[attr-defined] # noqa: F821
+			"next-to-last": cls._next_to_last  # type: ignore[attr-defined] # noqa: F821
+		}
+
+		try:
+			func = funcs[function_name]
+		except Exception:
```

Comment:
Just like above, `except Exception:` is an antipattern in Python unless you really mean “catch all non-critical errors”. In this case you only want to catch `KeyError` (I presume), so write that instead.

Additionally, if you write something like `except …: raise <ErrorType>(…)` in Python, the interpreter will assume that your code crashed while trying to handle the previous exception and will show this in the backtrace. If you are instead raising an exception because you caught another write `except … as exc: raise <ErrorType>(…) from exc` instead so that the interpret can show proper explanations of what happened in stack traces.